include/bitops.h

```diff
@@ -0,0 +1,224 @@
+//
+// Examples
+// ~~~~~~~~
+// uint8_t reg = 0;
+//
+// Setting:
+//   set_bits(reg, bit_0 | bit_1); // 0b0000'0011
+//   set_bits(reg, bit_8); // ASSERT FAIL (out of range)
+//   set_all_bits(reg); // 0b1111'1111
+//
+// Toggling:
+//   toggle_bits(reg, bit_4 | bit_5 | bit_6 | bit_7); // 0b0000'1111
+//   toggle_bits(reg, bit_8); // ASSERT FAIL (out of range)
+//   toggle_all_bits(reg); //  0b1111'0000
+//
+// Clearing:
+//   clear_bits(reg, bit_4 | bit_5); // 0b1100'0000
+//   clear_bits(reg, bit_8); // ASSERT FAIL (out of range)
+//
+// Checking:
+//   are_bits_set(reg, bit_6); // true
+//   are_bits_set(reg, bit_7); // true
+//   are_bits_set(reg, bit_8); // ASSERT FAIL (out of range)
+//   are_bits_set(reg, bit_6 | bit_7); // true (both need to be set)
+//   are_bits_set(reg, bit_5 | bit_6); // false (both aren't set)
+//   are_any_bits_set(reg, bit_5 | bit_6); // true (bit 6 is)
+//   are_bits_cleared(reg, bit_0 | bit_1 | bit_2 | bit_3); // true
+//
+
+enum BITS {
+	bit_0 = 1 << 0,
+	bit_1 = 1 << 1,
+	bit_2 = 1 << 2,
+	bit_3 = 1 << 3,
+	bit_4 = 1 << 4,
+	bit_5 = 1 << 5,
+	bit_6 = 1 << 6,
+	bit_7 = 1 << 7,
+	bit_8 = 1 << 8,
+	bit_9 = 1 << 9,
+	bit_10 = 1 << 10,
+	bit_11 = 1 << 11,
+	bit_12 = 1 << 12,
+	bit_13 = 1 << 13,
+	bit_14 = 1 << 14,
+	bit_15 = 1 << 15,
+	bit_16 = 1 << 16,
+	bit_17 = 1 << 17,
+	bit_18 = 1 << 18,
+	bit_19 = 1 << 19,
+	bit_20 = 1 << 20,
+	bit_21 = 1 << 21,
+	bit_22 = 1 << 22,
+	bit_23 = 1 << 23,
+	bit_24 = 1 << 24,
+	bit_25 = 1 << 25,
+	bit_26 = 1 << 26,
+	bit_27 = 1 << 27,
+	bit_28 = 1 << 28,
+	bit_29 = 1 << 29,
+	bit_30 = 1 << 30,
+	bit_31 = 1 << 31
+};
+
+// Is the bitmask within the maximum value of the register?
+template <typename T>
+static constexpr void validate_bit_width([[maybe_unused]] const T reg, const int bits)
+{
+	// ensure the register is unsigned
+	static_assert(std::is_unsigned<T>::value, "register must be unsigned");
+
+	// Ensure the bits fall within the type size
+	assert(static_cast<uint64_t>(bits) > 0);
+	assert(static_cast<int64_t>(bits) <=
+	       static_cast<int64_t>(std::numeric_limits<T>::max()));
+}
+
+// Set the indicated bits
+template <typename T>
+constexpr T set_bits_const(const T reg, const int bits)
+{
+	validate_bit_width(reg, bits);
+	return reg | static_cast<T>(bits);
+}
+template <typename T>
+constexpr void set_bits(T &reg, const int bits)
+{
+	reg = set_bits_const(reg, bits);
+}
+
+// Set all of the register's bits high
+//
+// It's often seen in code like this:
+//    unsigned long int flags = ~0; // [1]
+//    unsigned long int flags = 0xffffffff; // [2]
+//
+// But:
+//   [1] isn't portable because it relies on a two's-complement representation.
+//   [2] isn't portable because it assumes 32-bit ints.
+//
+// This function becomes self-documenting: it tells the reader that we intend to
+// use the bits in the register as opposed to being a plain old number or
+// counter.
+//
+template <typename T>
+constexpr T all_bits_set()
```

Comment:
Something about this name doesn't seem quite right to me.

I initially thought the function returned a bool because of the name.

Not sure what to change it to though.

Reply:
Yes; I also struggled with it. 
"All bits set" is meant to represent a type with all its bits enabled.  I wanted something that I could use when constructing a new variable. 
```c++
auto new_var = all_bits_set<unint16_t>();
```
Instead of:

``` c++
auto uint16_t new_var = 0;
flip_all_bits(new_var);
```

Reply:
> Not sure what to change it to though.

Yeah; whats the name of a type with all its bits set? It's different than the max value - because signed still is missing it's last bit being set.

Reply:
Just trying to brainstorm and throw stuff out there.

Reply:
I can't think of anything better to call it so may as well leave it as-is.